File: clients/ios/Lextures/Core/LMS/CourseFileLogic.swift
import Foundation

enum CourseFileLogic {
    static func contentPath(courseCode: String, source: CourseFileContentSource) -> String {
        let encoded = LMSAPI.encodePath(courseCode)
        switch source {
        case .fileManager(let itemId):
            return "/api/v1/courses/\(encoded)/files/items/\(LMSAPI.encodePath(itemId))/content"
        case .courseFile(let fileId):
            return "/api/v1/courses/\(encoded)/course-files/\(LMSAPI.encodePath(fileId))/content"
        }
    }

    static func previewPath(courseCode: String, itemId: String) -> String {
        let encoded = LMSAPI.encodePath(courseCode)
        return "/api/v1/courses/\(encoded)/files/items/\(LMSAPI.encodePath(itemId))/preview"
    }

    static func downloadKey(courseCode: String, target: FilePreviewTarget) -> String {
        "download:\(courseCode):\(target.sourceKey)"
    }

    static func courseFilesCacheKey(courseCode: String, folderId: String?) -> String {
        if let folderId, !folderId.isEmpty {
            return "course:\(courseCode):files:folder:\(folderId)"
        }
        return "course:\(courseCode):files:root"
    }

    static func previewKind(mimeType: String?, fileName: String) -> FilePreviewKind {
        let mime = (mimeType ?? "").lowercased()
        let ext = (fileName as NSString).pathExtension.lowercased()

        if mime.hasPrefix("image/") || ["png", "jpg", "jpeg", "gif", "webp", "heic", "svg"].contains(ext) {
            return .image
        }
        if mime == "application/pdf" || ext == "pdf" {
            return .pdf
        }
        if mime.hasPrefix("audio/") || ["mp3", "wav", "m4a", "aac", "ogg"].contains(ext) {
            return .audio
        }
        if mime.hasPrefix("video/") || ["mp4", "mov", "webm", "m4v"].contains(ext) {
            return .video
        }
        return .downloadOnly
    }

    static func guessMimeType(from fileName: String) -> String? {
        let ext = (fileName as NSString).pathExtension.lowercased()
        switch ext {
        case "pdf": return "application/pdf"
        case "png": return "image/png"
        case "jpg", "jpeg": return "image/jpeg"
        case "gif": return "image/gif"
        case "webp": return "image/webp"
        case "mp4": return "video/mp4"
        case "mov": return "video/quicktime"
        case "mp3": return "audio/mpeg"
        case "wav": return "audio/wav"
        case "m4a": return "audio/mp4"
        default: return nil
        }
    }

    static func formatBytes(_ bytes: Int64) -> String {
        ByteCountFormatter.string(fromByteCount: bytes, countStyle: .file)
    }

    static func systemImage(for kind: FilePreviewKind) -> String {
        switch kind {
        case .image: return "photo"
        case .pdf: return "doc.richtext"
        case .audio: return "waveform"
        case .video: return "play.rectangle"
        case .downloadOnly: return "doc"
        }
    }

    static func listIcon(isFolder: Bool, fileName: String, mimeType: String) -> String {
        if isFolder { return "folder.fill" }
        return systemImage(for: previewKind(mimeType: mimeType, fileName: fileName))
    }

    static func accessibilityLabel(name: String, isFolder: Bool, mimeType: String?, byteSize: Int64?) -> String {
        var parts = [isFolder ? "Folder" : "File", name]
        if let mimeType, !mimeType.isEmpty, !isFolder {
            parts.append(mimeType)
        }
        if let byteSize, byteSize > 0 {
            parts.append(formatBytes(byteSize))
        }
        return parts.joined(separator: ", ")
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/CourseFileLogic.kt
package com.lextures.android.core.lms

import java.net.URLEncoder
import java.text.DecimalFormat

object CourseFileLogic {
    fun encodePath(value: String): String = URLEncoder.encode(value, "UTF-8").replace("+", "%20")

    fun contentPath(courseCode: String, source: CourseFileContentSource, sourceId: String): String =
        when (source) {
            CourseFileContentSource.FileManager ->
                "/api/v1/courses/${encodePath(courseCode)}/files/items/${encodePath(sourceId)}/content"
            CourseFileContentSource.CourseFile ->
                "/api/v1/courses/${encodePath(courseCode)}/course-files/${encodePath(sourceId)}/content"
        }

    fun previewPath(courseCode: String, itemId: String): String =
        "/api/v1/courses/${encodePath(courseCode)}/files/items/${encodePath(itemId)}/preview"

    fun downloadKey(courseCode: String, target: FilePreviewTarget): String =
        "download:$courseCode:${target.id}"

    fun courseFilesCacheKey(courseCode: String, folderId: String?): String =
        if (!folderId.isNullOrEmpty()) "course:$courseCode:files:folder:$folderId"
        else "course:$courseCode:files:root"

    fun previewKind(mimeType: String?, fileName: String): FilePreviewKind {
        val mime = mimeType.orEmpty().lowercase()
        val ext = fileName.substringAfterLast('.', "").lowercase()
        return when {
            mime.startsWith("image/") || ext in setOf("png", "jpg", "jpeg", "gif", "webp", "heic", "svg") ->
                FilePreviewKind.Image
            mime == "application/pdf" || ext == "pdf" -> FilePreviewKind.Pdf
            mime.startsWith("audio/") || ext in setOf("mp3", "wav", "m4a", "aac", "ogg") -> FilePreviewKind.Audio
            mime.startsWith("video/") || ext in setOf("mp4", "mov", "webm", "m4v") -> FilePreviewKind.Video
            else -> FilePreviewKind.DownloadOnly
        }
    }

    fun guessMimeType(fileName: String): String? {
        return when (fileName.substringAfterLast('.', "").lowercase()) {
            "pdf" -> "application/pdf"
            "png" -> "image/png"
            "jpg", "jpeg" -> "image/jpeg"
            "gif" -> "image/gif"
            "webp" -> "image/webp"
            "mp4" -> "video/mp4"
            "mov" -> "video/quicktime"
            "mp3" -> "audio/mpeg"
            "wav" -> "audio/wav"
            "m4a" -> "audio/mp4"
            else -> null
        }
    }

    fun formatBytes(bytes: Long): String {
        if (bytes <= 0) return "0 B"
        val units = arrayOf("B", "KB", "MB", "GB")
        var value = bytes.toDouble()
        var unit = 0
        while (value >= 1024 && unit < units.lastIndex) {
            value /= 1024
            unit++
        }
        return DecimalFormat("#.#").format(value) + " " + units[unit]
    }
}
